### ISSUE TITLE
[docdb] load balance single shard tables

### ISSUE THREAD

**Issue by jaki** [yugabyte/yugabyte-db#4407]
Load balancing doesn't seem to work on single shard (single tablet-peer) tables.

YSQL hash:

```sh
./bin/yb-ctl create --ysql_num_shards_per_tserver 1 --rf 1
for i in $(seq 1 20); do
  ./bin/ysqlsh --command "CREATE TABLE t$i (i int)"
done
./bin/yb-ctl add_node
```

YCQL hash:

```sh
./bin/yb-ctl create --num_shards_per_tserver 1 --rf 1
./bin/cqlsh --execute "CREATE KEYSPACE k"
for i in $(seq 1 20); do
  ./bin/cqlsh --execute "CREATE TABLE t$i (i int PRIMARY KEY)" --keyspace k
done
./bin/yb-ctl add_node
```

YSQL range:

```sh
./bin/yb-ctl create --rf 1
for i in $(seq 1 20); do
  ./bin/ysqlsh --command "CREATE TABLE t$i (i int, PRIMARY KEY (i ASC))"
done
./bin/yb-ctl add_node
```

YSQL hash, rf 3:

```gdb
# This is a gdb script named script.gdb
# break on the line that looks like
#   LOG(INFO) << "Creating partitions with num_tablets: " << num_tablets;
delete
break src/yb/common/partition.cc:390
  commands
    print num_tablets = 1
    continue
  end
continue
```

```sh
./bin/yb-ctl create --rf 3
gdb -p <master_leader_pid> --eval-commands 'source script.gdb'
for i in $(seq 1 20); do
  ./bin/ysqlsh --command "CREATE TABLE t$i (i int)"
done
./bin/yb-ctl add_node
```

After load balancing is finished, check http://127.0.0.1:7000/tablet-servers to see that **all** user tablets have not moved.  For the rf 1 cases, they are still on the first tserver (127.0.0.1) rather than the new tserver (127.0.0.2).  For the rf 3 case, the raft group is still the first three tservers and doesn't include the new fourth tserver.

**Comment by jaki** [comment]
Furthermore, replication factor is not met when you remove nodes.  Continuing off the rf 3 example above, running

```sh
./bin/yb-ctl remove_node 1
```

causes the tablets whose leader was tserver 1 to change leaders, as expected.  However, the raft configuration still seems to show that the group is composed of the original three tservers (127.0.0.{1,2,3}), not involving the new fourth tserver.  The new fourth tserver still hasn't acquired any load, and the load balancer seems to be continually active.  The master leader continually shows log messages like

```
W0506 15:23:13.277345  9939 cluster_balance.cc:227] Skipping load balancing 000030a9000030008000000000004012: Leader not ready to serve requests. (../../src/yb/master/cluster_balance_util.h:286): Master leader has not yet received heartbeat from ts 2845aec46e134f2c871c84ab4d256371, either master just became leader or a network partition.
```

**Comment by jaki** [comment]
This is on commit 5ff03f067500584971c522a44c1c4a80213b7c56, centos, debug, ninja build.

**Comment by bmatican** [comment]
@rahuldesirazu do you think you can take a look at this?

@jaki for the 2nd part, when you remove a node, that's expected, as `yb-ctl remove_node` still just does a kill, so we'll wait out the default 15m of follower unavailability until raft kicks out the dead node, until the LB is back online.

**Comment by jaki** [comment]
This also affects YSQL colocated databases, whose colocation tablets are single shards.

**Comment by jaki** [comment]
One thing I noticed is that `tablet_meta.is_over_replicated` in `src/yb/master/cluster_balance_util.h` never becomes true for these single shard tablets but (sometimes) does become true for the other cases, like the transaction status tablet and tablets of multi-partitioned tables.

**Comment by jaki** [comment]
@bmatican, 15 minutes have passed, and the load balancer is now idle.  tserver 4 is now in use!  But it still doesn't have any leaders (but that may be expected).  http://127.0.0.1:7000/tablet-servers shows

| ip | user tablet-peers | leaders |
| --- | --- | --- |
| 127.0.0.2 | 20 | 11 |
| 127.0.0.3 | 20 | 9 |
| 127.0.0.4 | 20 | 0 |

**Comment by bmatican** [comment]
@jaki Yeah, for leaders it's a bit iffy, cus we don't do a global pass, atm.

So from a per-table perspective, there's just 1 tablet, so the difference in leader counts per TS for this table is at most 1, which is the threshold we stop at normally (imagine X vs X+1, per table, there's no point to move further).

**Comment by bmatican** [comment]
Actually, thinking some more about this and seeing one of @jaki 's other diffs, I realized that this is the same for leaders as it is for data.

Fundamentally, the LB operates per tablet atm. If a table has just has one tablet, with say 3 peers, and you have more tservers, say, 4, then the distribution will be 1110. The variance between highest and lowest loaded will be 1, which is the cutoff point for the LB, as it cannot guarantee any local (per table) moves will accomplish any more balance.

This will be addressed once we tackle #3040 and #3079

**Comment by hulien22** [comment]
Closing since we have 83bf39c from #3079. 

Note that we still haven't done #3040 for leader moves yet, but that is not as critical.
